### ISSUE TITLE
Unique token filter results in tokens with misplaced positions

### ISSUE THREAD

**Issue by shuaibiyy** [elastic/elasticsearch#35411]
**Elasticsearch version** (`bin/elasticsearch --version`):

Version: 6.4.2, Build: oss/tar/04711c2/2018-09-26T13:34:09.098244Z, JVM: 1.8.0_181

**Plugins installed**: None

**JVM version** (`java -version`): java version "1.8.0_181"

**OS version** (`uname -a` if on a Unix-like system):

Ubuntu 16.04.5 LTS Linux 4.4.0-1069-aws #79-Ubuntu SMP Mon Sep 24 15:01:41 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
Unique token filter assigns incorrect positions to tokens that have had their first character deduplicated.

**Steps to reproduce**:

*Create index:*
```
curl -XPUT localhost:9200/test
```

*Expected:* Analysis of a text `los angeles leadership academy` using the `edge_ngram` and `unique` token filters should result in unique tokens sized at the `max_gram` (default 2) of the `edge_ngram` filter and span the positions of the 4 words.
```
curl -XPOST localhost:9200/test/_analyze -H 'Content-Type:application/json' -d '{
    "tokenizer": "standard",
    "text": "los angeles leadership academy",
    "filter": ["edge_ngram", "unique"]
}'
```

*Actual:*
The tokens `l, lo, a, an, le, ac` are produced. However, the tokens `le` and `ac` are incorrectly placed at position 1 due to the first letters in their respective words getting deduplicated by the `unique` token filter. See below:
```
{
  "tokens": [
    {
      "token": "l",
      "start_offset": 0,
      "end_offset": 3,
      "type": "<ALPHANUM>",
      "position": 0
    },
    {
      "token": "lo",
      "start_offset": 0,
      "end_offset": 3,
      "type": "<ALPHANUM>",
      "position": 0
    },
    {
      "token": "a",
      "start_offset": 4,
      "end_offset": 11,
      "type": "<ALPHANUM>",
      "position": 1
    },
    {
      "token": "an",
      "start_offset": 4,
      "end_offset": 11,
      "type": "<ALPHANUM>",
      "position": 1
    },
    {
      "token": "le",
      "start_offset": 12,
      "end_offset": 22,
      "type": "<ALPHANUM>",
      "position": 1
    },
    {
      "token": "ac",
      "start_offset": 23,
      "end_offset": 30,
      "type": "<ALPHANUM>",
      "position": 1
    }
  ]
}

```

A different text, `los angeles city hall` for example, would produce the tokens `l, lo, a, an, c, ci, h, ha` that correctly span positions 0 to 3.

**Comment by romseygeek** [comment]
Looks like UniqueTokenFilter isn't keeping track of the position increments of dropped tokens.  We can re-implement it as a FilteringTokenFilter which should deal with all the book-keeping for us.

**Comment by shuaibiyy** [comment]
Hi @romseygeek I attempted a fix in https://github.com/elastic/elasticsearch/issues/35411

**Comment by DerpgonCz** [comment]
Hi, this is kind of a breaking bug. There is no workaround that I found. As far as I tested it, it is not fixed in ES 8.13 yet. Any updates on this?

**Comment by DerpgonCz** [comment]
@benwtrent Actual GOAT, thanks a lot!
